docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+# ADR 004: Coroutine-Based Effect System for Consensus
+
+## Changelog
+
+* 18-03-2025: Initial version
+
+## Status
+
+Accepted
+
+## Context
+
+The Malachite consensus engine needs to interact with its environment (network, storage, cryptography, application logic) at specific points during execution.
+
+Traditional approaches to handling these interactions include:
+
+1. **Callback-based designs** - Provide callback interfaces that must be implemented by users
+2. **Trait-based polymorphism** - Define traits that users implement to provide required functionality
+3. **Message-passing architectures** - Define protocols for communication between consensus and external components
+
+We needed a design that would:
+- Maintain a clear separation between the consensus algorithm and its environment
+- Keep the consensus code linear and readable despite external interactions
+- Support both synchronous and asynchronous operations
+- Facilitate testing by making effects explicit and mockable
+- Allow different execution environments (sync/async runtimes, actor systems, etc.)
```

Comment:
Is this the same as 

> - Support both synchronous and asynchronous operations

?

Reply:
No, the terms sync/async are unfortunately a bit overloaded here.

For _effects_:
- Sync: the effect must be performed immediately and the coroutine resumed with a result value
- Async: the effect can be performed in the background, the coroutine resumed without a value, and the result value provided at a later point via an input.

For the execution environment:
- Sync: the environment performs blocking syscalls
- Async: the environment uses an async runtime like Tokio without blocking syscalls

Reply:
So operations are the latter?

Reply:
clarified in text